### PR TITLE
8341834: C2 compilation fails with "bad AD file" due to Replicate

### PR DESCRIPTION
**Pull request by rwestrel**
Superword creates a `Replicate` node at a `ConvL2I` node and uses the
type of the result of the `ConvL2I` to pick the type of the
`Replicate` instead of the type of the input to the `ConvL2I`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341834](https://bugs.openjdk.org/browse/JDK-8341834): C2 compilation fails with "bad AD file" due to Replicate (**Bug** - P3)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21660/head:pull/21660` \
`$ git checkout pull/21660`

Update a local copy of the PR: \
`$ git checkout pull/21660` \
`$ git pull https://git.openjdk.org/jdk.git pull/21660/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21660`

View PR using the GUI difftool: \
`$ git pr show -t 21660`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21660.diff">https://git.openjdk.org/jdk/pull/21660.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21660#issuecomment-2431308174)
</details>

### CODE REVIEW

**Comment by TobiHartmann** [comment]
Is this a regression from [JDK-8332163](https://bugs.openjdk.org/browse/JDK-8332163) or [JDK-8248830](https://bugs.openjdk.org/browse/JDK-8248830)?

**Comment by eme64** [comment]
Hi Roland,

Isn't it a little strange that we have a Replicate before a ConvL2I pack? That means that the ConvL2I did not common, even though they have the same inputs. I guess that is due to the `type` not being identical - a rather rare case.

```
 Pack: 8
    0:  739  ConvL2I  === _ 742  [[ 738 ]]  #int:2..99:www !orig=612,467,407,[138],[156] !jvms: Test4::test @ bci:41 (line 11)
    1:  741  ConvL2I  === _ 742  [[ 740 ]]  #int:2..198:www !orig=613,468,440
 Pack: 9
    0:  755  ConvL2I  === _ 756  [[ 738 ]]  #int:2..109:www !orig=594,458,[399],137 !jvms: Test4::test @ bci:37 (line 10)
    1:  754  ConvL2I  === _ 756  [[ 740 ]]  #int:2..208:www !orig=591,514
```

I played around with the test case as well, reducing it further:

`./java -XX:CompileCommand=quiet -XX:CompileCommand=compileonly,Test4::test -Xcomp -XX:+TraceSuperWord -XX:+TraceNewVectors -XX:UseAVX=2 Test4.java`

```
public class Test4 {
    public static long val = 0;

    public static void test(int x) {
        x = Math.max(0, Math.min(10, x)); // type 0..10
        short a[] = new short[500];

        for (long l = 0; l < 100; l++) {
            val = l + x; // store seems required, hmm
            int y = (int)val;
            int z = (int)l; // this becomes multiple ConvL2I
            a[z] = (short)(z - y);
        }
    }

    public static void main(String[] args) {
        Math.min(0, 1);
        Math.max(0, 1);
        test(0);
    }
}
```

**Comment by eme64** [comment]
The fix seems correct, but I fear that maybe older versions could be affected, it would just be very difficult to create that `ConvL2I` pack.

**Comment by eme64** [comment]
I would add a comment in the VM code. The one-liner is a little dense to read.

Basically the issue is that `velt_basic_type(p0)` gives us the output type of `p0`. But what we need is the input type of `p0` - this is what we are replicating for.

Now I'm wondering if there are any other `p0` nodes that have diverging input/output type?

File: src/hotspot/share/opto/superwordVTransformBuilder.cpp [hunk 1]
```diff
@@ -228,7 +228,7 @@ VTransformNode* SuperWordVTransformBuilder::get_or_make_vtnode_vector_input_at_i
       return shift_count;
     } else {
       // Replicate the scalar same_input to every vector element.
-      BasicType element_type = _vloop_analyzer.types().velt_basic_type(p0);
+      BasicType element_type = p0->is_Convert() ? p0->in(1)->bottom_type()->basic_type() : _vloop_analyzer.types().velt_basic_type(p0);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/superwordVTransformBuilder.cpp:231]: What vectors are generated (or not) with this change? The array in the test ins `int[]` but the element_type will be Long now. Will it bailout vectorization?

Comment (vnkozlov) [reply src/hotspot/share/opto/superwordVTransformBuilder.cpp:231]: Thank you for explanation. Yes, to have comment would be nice.

Comment (rwestrel) [reply src/hotspot/share/opto/superwordVTransformBuilder.cpp]: I added a comment. Does it look ok to you?

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by eme64 — APPROVED** [review]
Nice, thanks for the added comments!

Do you know what JDK versions are affected?

**Comment by rwestrel** [comment]
> Do you know what JDK versions are affected?

The failure doesn't reproduce with jdk21u. But that seems to be because we need JDK-8326139 (and JDK-8331575) for the bug to show up.

**Comment by shqking** [comment]
Hi @rwestrel 

My JBS account is inactive recently. Hence I'd like to report the bug here.

I encountered the following error with `-XX:MaxVectorSize=8` on both AArch64 and x86_64.
Could you help take a look at this issue? Thanks.

Test command:
```
make test JTREG="VM_OPTIONS=-XX:MaxVectorSize=8" TEST=test/hotspot/jtreg/compiler/vectorization/TestReplicateAtConv.java
```

Error message:
```
CompileCommand: compileonly TestReplicateAtConv.test bool compileonly = true            
#                                                                                       
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/tmp/jdk-dev/src/hotspot/share/opto/type.cpp:2499), pid=1424540, tid=1424557
#  assert(Matcher::vector_size_supported(elem_bt, length)) failed: length in range
#
# JRE version: OpenJDK Runtime Environment (24.0) (fastdebug build 24-internal-git-63c19d3db58)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 24-internal-git-63c19d3db58, compiled mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# V  [libjvm.so+0x17bca30]  TypeVect::make(BasicType, unsigned int, bool)+0x150
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /tmp/jdk-build/test-support/jtreg_test_hotspot_jtreg_compiler_vectorization_TestReplicateAtConv_java/scratch/0/core.1424540)
#
# An error report file with more information is saved as:
# /tmp/jdk-build/test-support/jtreg_test_hotspot_jtreg_compiler_vectorization_TestReplicateAtConv_java/scratch/0/hs_err_pid1424540.log
#
# Compiler replay data is saved as:
# /tmp/jdk-build/test-support/jtreg_test_hotspot_jtreg_compiler_vectorization_TestReplicateAtConv_java/scratch/0/replay_pid1424540.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
```

**Comment by rwestrel** [comment]
@shqking thanks for the report. I filed https://bugs.openjdk.org/browse/JDK-8343747

**Pull request closed without merging**